platform/eventing/listener.go

```diff
@@ -0,0 +1,129 @@
+package eventing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/cloudevents/sdk-go/v2/client"
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	cloudeventsnats "github.com/cloudevents/sdk-go/v2/protocol/nats"
+	"github.com/fewlinesco/go-pkg/platform/database"
+	"github.com/fewlinesco/go-pkg/platform/monitoring"
+	"go.opencensus.io/trace"
+)
+
+var (
+	ErrConsumerEventCanNotBePersisted = errors.New("the event could not be persisted in the database")
+)
+
+type listenerSubject string
+
+// Listener subscribes to subjects/queues in the broker and persists them in a local database
+type Listener struct {
+	URL      string
+	subjects []string
+	db       *database.DB
+	logger   *log.Logger
+
+	stop               chan bool
+	maxNumberOfRetries int
+}
+
+// NewListener creates a new config for a listener
+func NewListener(URL string, subjects []string, db *database.DB, logger *log.Logger) *Listener {
+	return &Listener{
+		URL:      URL,
+		subjects: subjects,
+		db:       db,
+		logger:   logger,
+
+		stop:               make(chan bool, 0),
+		maxNumberOfRetries: 5,
+	}
+}
+
+func (listener *Listener) Stop() {
+	listener.stop <- true
+}
+
+// Start starts a listener for the provided subjects
+// It will save all events for the subjects in the DB
+func (listener *Listener) Start() error {
+	shutdown := make(chan error, 0)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	for _, subject := range listener.subjects {
+		go func(ctx context.Context) {
+			ctx = context.WithValue(ctx, listenerSubject(subject), subject)
+
+			listener.logger.Printf("consumer started for: %s", subject)
+
+			if err := startReceiver(ctx, listener, subject); err != nil {
+				shutdown <- fmt.Errorf("receiver '%s': %v", subject, err)
+			}
+		}(ctx)
+	}
+
+	select {
+	case <-listener.stop:
+		cancel()
+		return nil
+	case err := <-shutdown:
+		cancel()
+		return fmt.Errorf("all listeners stopped: %v", err)
+	}
+}
+
+func startReceiver(ctx context.Context, listener *Listener, subject string) error {
+	natsConsumer, err := cloudeventsnats.NewConsumer(listener.URL, subject, cloudeventsnats.NatsOptions())
+	if err != nil {
+		return fmt.Errorf("failed to create nats consumer, %v", err)
+	}
+
+	natsClient, err := client.New(natsConsumer)
+	if err != nil {
+		return fmt.Errorf("failed to create client, %v", err)
+	}
+
+	go func(ctx context.Context) {
+		select {
+		case <-ctx.Done():
+			natsConsumer.Close(ctx)
+		}
+	}(ctx)
+
+	// do we need to keep a timeframe?
```

Comment:
If we don't I suggest remove the comment and if we want, I would suggest creating a ticket :)